### PR TITLE
build: introduce PKG_BUILD_FLAGS

### PR DESCRIPTION
**Pull request by dhewg**
PKG_BUILD_FLAGS is a new variable for package Makefiles similar to PKG_FLAGS. It's a whitespace separated list of flags to control various aspects of how a package is build.
    
The build system and/or .config defines the default for each, but every package has the means to override it. Using $flagname enables a flag, no-$flagname disables it.

This refactors PKG_IREMAP and PKG_USE_MIPS16 to such flags, then adds "gc-sections" and "lto" on top.

With that in place we can do nice things like playing with the added experimental knobs CONFIG_USE_GC_SECTIONS and CONFIG_USE_LTO to set different defaults treewide.

Some opt-outs here are already in place, found by @robimarko and myself.
For the package feeds there're some too, found here:
https://github.com/dhewg/packages/commits/mold

This is in part for #11895, where I added two new flags.

### CODE REVIEW

**Comment by Ansuel** [comment]
Should we really care to have the 'no-' logic and an additional function instead of just adding specific filter? Isn't that overkill for just a bunch of options?

**Comment by dhewg** [comment]
That's by design. Currently we only have per-package bool flags.
This adds the possibility for flipping the default, which is what the two config knobs do.

Like, grub2 has the `no-lto` flag because we know that it fails at life. It has that even though it doesn't enable lto by itself.
But then we have the experimental option to enable lto tree-wide. So all packages with grub2 as one exception are built with lto. In other words: a no-foo flag overwrites a global foo=true. Without the `no-lto` flag it would fail to build if attempting lto tree-wide.

Does that make sense?

**Comment by dhewg** [comment]
Apart from that I think it's a nice cleanup already, it removes some PKG_* vars, and there're more candidates (pic, all of hardening.mk), with just a single one where you just add the desired flags.

**Comment by Ansuel** [comment]
@dhewg nooo I think you got this wrong. To me it seems overkill adding the pkg_build_flag function. I would honestly use filter where needed and it may be even more readable. 

Also parallel is a valid candidate for pkg_build_flags conversion

**Comment by dhewg** [comment]
It's just
`if (no-$flagname in $flags) return 0 else if ($flag in $flags) return 1 else return $default`.
I mean I agree it's no beauty, but that's just make syntax, not my primary weapon of choice. There're lots of worse offenders than that oneliner.

Random example:
`SUBTARGET:=$(strip $(foreach subdir,$(patsubst $(PLATFORM_DIR)/%/target.mk,%,$(wildcard $(PLATFORM_DIR)/*/target.mk)),$(if $(CONFIG_TARGET_$(call target_conf,$(BOARD)_$(subdir))),$(subdir))))`
Not exactly easy to digest.

The mold PR add two new flags, there're other candidates. If I make unfold that check for positive and/or negative for all flags, the next guy will see the reoccurring pattern and refactor that into a function like I have now.

**Comment by dhewg** [comment]
I'll stare at it some more tomorrow, maybe I can think of something to less it less ugly.
Suggestions more than welcome ;)

**Comment by Ansuel** [comment]
To me it seems more readable with filter and the option directly.

Il Dom 5 Feb 2023, 00:09 dhewg ***@***.***> ha scritto:

**Comment by robimarko** [comment]
I kind of like the `PKG_BUILD_FLAGS` as a one-stop place instead of adding more knobs

**Review by hauke — APPROVED** [review]

File: include/package.mk [hunk 1]
```diff
@@ -30,7 +29,15 @@ ifdef CONFIG_USE_MIPS16
...
-ifeq ($(strip $(PKG_IREMAP)),1)
+
+PKG_BUILD_FLAGS?=
+
+# $1=flagname, $2=default (0/1)
+define pkg_build_flag
+$(if $(filter no-$(1),$(PKG_BUILD_FLAGS)),0,$(if $(filter $(1),$(PKG_BUILD_FLAGS)),1,$(2)))
+endef
+
+ifeq ($(call pkg_build_flag,iremap,1),1)
```
Comment (Ansuel) [inline include/package.mk:38]: I'm ok with the changes... Only i think having this amount of freedom may introduce some ""problems"" in the future...

Also to me something like

```
ifeq ($(filter no-iremap, $(PKG_BUILD_fLAGS)),)
```

Looks more direct and clear than a function and a default value...

The idea of these particular flag is to deviate from default configuration in case of problems and ideally they should be there only to workaround unfixable problems.

Ideally each package should use the same flags and if possible apply some fix to follow that.

Giving the possibility to support both no- and normal option CAN result in giving more flexibility and spending less time in trying to make the package more supported...

So IMHO providing only the specific workaround is better than giving the option to follow that or not.

In short I would ask you to drop the define small function and just convert to PKG_BUILD_FLAGS providing directly the supported option.

Free to use whatever syntax you prefer to make more clear what the ifeq does
Example (for the same solution)

```
ifneq ($(filter no-iremap, $(PKG_BUILD_fLAGS)),no-iremap)
```

Comment (dhewg) [reply include/package.mk:38]: For this specific flag, note that you get the very same result with `PKG_IREMAP:=0` right now.

As for the function, I explicitly wanted a way to:
* let a package set a flag even if the global default doesn't set it
* let a package unset a flag even if the global default sets it

For some flags the default is set, for others unset. Some defaults can be toggled via .config. The function covers all permutations.
If I just check only for $flag and not no-$flag (or the other way around) it doesn't work anymore as intended.
So you really want me to unfold that function for every flag?

**Comment by dhewg** [comment]
@Ansuel maybe we're talking past each other? Does this help?

|old var|new flag|default if|in use|notes|
|---|---|---|---|---|
|PKG_IREMAP:=1|iremap|always|n|CONFIG_REPRODUCIBLE_DEBUG_INFO controls its behaviour|
|PKG_IREMAP:=0|no-iremap||n|may be useful locally for debugging?|
|PKG_USE_MIPS16:=1|mips16|CONFIG_USE_MIPS16=y|n|noop if CONFIG_USE_MIPS16=n|
|PKG_USE_MIPS16:=0|no-mips16|CONFIG_USE_MIPS16=n|y|overrides CONFIG_USE_MIPS16=y|
|-|gc-sections|CONFIG_USE_GC_SECTIONS=y|y|C/LFLAGS used manually and inconsistently before|
|-|no-gc-sections|CONFIG_USE_GC_SECTIONS=n|n|overrides CONFIG_USE_GC_SECTIONS=y|
|-|lto|CONFIG_USE_LTO=y|y|C/LFLAGS used manually and inconsistently before|
|-|no-lto|CONFIG_USE_LTO=n|y|overrides CONFIG_USE_LTO=y, set if LTO breaks stuff|
|-|mold|CONFIG_USE_MOLD=y|n|noop if CONFIG_USE_MOLD=n|
|-|no-mold|CONFIG_USE_MOLD=n|y|overrides CONFIG_USE_MOLD=y, set if linking with mold doesn't work|

**Comment by dhewg** [comment]
Thanks for all the feedback, but to be honest: I don't get it.

If we somehow get to the goal of "common option for every package" there wouldn't be a need for any PKG_* var nor any flags? But I don't see that happening, ever?

Packages are not supposed get a "no-mold" because someone didn't like it. It's for cases where mold doesn't support a feature the package is using. Right now those are linker scripts. Without the opt-out there's no point in adding mold support, because we already know that it won't build.

And the flags are still part of the review process. If we assume maintainers or contributers get stuff in just because and without any technical reason, I can think of many other ways to break stuff and make maintanance harder.

What's redundant is adding "iremap", "mips16" or "mold" as they're a noop at this point, but what's the difference to PKG_IREMAP:=1 or PKG_USE_MIPS16:=1 ?

If the consent is that we don't want any no-$flag, that disqualifies the CONFIG_USE_LTO and CONFIG_USE_GC_SECTIONS patches, because that too breaks the build. I don't expect those flags to get enabled by default anyway, but it's still a quick and easy way to inspect size differences to fish out candidates to add "lto" or "gc-sections" to.

And if we're at this point we might as well close this. Which is fine. But then I'm back to PKG_NO_MOLD and still wonder what the point is.

**Comment by dhewg** [comment]
Re-reading your last commit, it sounds like you may have a wrong impression about the mold flag.
As is, the mold PR implements it as global opt-in. Without CONFIG_USE_MOLD=y adding `mold` to the flags doesn't do anything. Like I mentioned on that table above. So it already is the second scenario you described.

Does that resolve it? Or should I close this and move on?

**Comment by dhewg** [comment]
I'll take that as a 'no' then?
FWIW, I still think this flags thing is a really nice addition and cleanup. I didn't expect this to be controversial at all.

**Comment by dhewg** [comment]
I can only put it in so many words, if the message doesn't get across what's the point in repeating it?

But okay, how about this attempt:
There is no `mold` flag. Just like there is no `mips16` flag.
While the `pkg_build_flag` function is used for those two flags too (and that _does_ check for enabling _and_ disabling entries), the calls are wrapped in a `ifdef CONFIG_USE_[MIPS16|MOLD]`.
Which makes `mold` and `mips16` a no-op.
It has the same effect as adding `iwantapony` to the flags: it does exactly nothing ;)

**Comment by dhewg** [comment]
Hmm sure, but I guess that makes it even more ugly?

How about we document only valid flags instead?
I propose to move this file https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md here (also under packages/), and then document no-mips16, gc-sections, no-gc-sections, lto, no-lto, no-mold in there?

**Comment by Ansuel** [comment]
> Hmm sure, but I guess that makes it even more ugly?

mhhh why? Seems sensible to me to reject any kind of `iwantapony ` case

Yes Documenting it is good and should be added to the list to do. If we really can't follow that path, the wiki entry for the Makefile guide should be updated for sure.

**Comment by dhewg** [comment]
So not just the two no-ops, something generic to check for unknown flags.

Like this?
```
__unknown_flags=$(filter-out no-iremap no-mips16 gc-sections no-gc-sections lto no-lto no-mold,$(PKG_BUILD_FLAGS))
ifneq ($(__unknown_flags),)
  $(error unknown PKG_BUILD_FLAGS: $(__unknown_flags))
endif
```

**Comment by Ansuel** [comment]
@dhewg yes in theory that way we should also have a central place where we have each flag listed and supported. What do you think? (also a good place to spam some comments about them)

**Comment by anomeome** [comment]
Both knobs enabled:

```
make package/libs/openssl/{clean,compile} V=sc

/home/kc/wrtpac/source/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-12.2.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/12.2.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: /home/kc/wrtpac/source/tmp/ccIfc6nY.ltrans0.ltrans.o:(.data.rel.ro.legacy_digests+0x18): undefined reference to `ossl_ripemd160_functions'
collect2: error: ld returned 1 exit status
make[4]: *** [Makefile:18693: providers/legacy.so] Error 1
make[4]: Leaving directory '/home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/openssl-3.0.8'
make[3]: *** [Makefile:1572: build_sw] Error 2
make[3]: Leaving directory '/home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/openssl-3.0.8'
make[2]: *** [Makefile:400: /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/openssl-3.0.8/.built] Error 2
make[2]: Leaving directory '/home/kc/wrtpac/source/package/libs/openssl'
time: package/libs/openssl/compile#102.03#8.85#48.33
    ERROR: package/libs/openssl failed to build.
make[1]: *** [package/Makefile:116: package/libs/openssl/compile] Error 1
make[1]: Leaving directory '/home/kc/wrtpac/source'
make: *** [/home/kc/wrtpac/source/include/toplevel.mk:231: package/libs/openssl/compile] Error 2

```

**Comment by dhewg** [comment]
Works for me with
```
#
# Build Options
#
CONFIG_OPENSSL_OPTIMIZE_SPEED=y
CONFIG_OPENSSL_WITH_ASM=y
# CONFIG_OPENSSL_WITH_DEPRECATED is not set
CONFIG_OPENSSL_NO_DEPRECATED=y
CONFIG_OPENSSL_WITH_ERROR_MESSAGES=y

#
# Protocol Support
#
CONFIG_OPENSSL_WITH_TLS13=y
# CONFIG_OPENSSL_WITH_DTLS is not set
# CONFIG_OPENSSL_WITH_NPN is not set
CONFIG_OPENSSL_WITH_SRP=y
CONFIG_OPENSSL_WITH_CMS=y

#
# Algorithm Selection
#
# CONFIG_OPENSSL_WITH_EC2M is not set
CONFIG_OPENSSL_WITH_CHACHA_POLY1305=y
CONFIG_OPENSSL_PREFER_CHACHA_OVER_GCM=y
CONFIG_OPENSSL_WITH_PSK=y

#
# Less commonly used build options
#
# CONFIG_OPENSSL_WITH_ARIA is not set
# CONFIG_OPENSSL_WITH_CAMELLIA is not set
# CONFIG_OPENSSL_WITH_IDEA is not set
# CONFIG_OPENSSL_WITH_SEED is not set
# CONFIG_OPENSSL_WITH_SM234 is not set
# CONFIG_OPENSSL_WITH_BLAKE2 is not set
# CONFIG_OPENSSL_WITH_MDC2 is not set
# CONFIG_OPENSSL_WITH_WHIRLPOOL is not set
# CONFIG_OPENSSL_WITH_COMPRESSION is not set
# CONFIG_OPENSSL_WITH_RFC3779 is not set

#
# Engine/Hardware Support
#
# CONFIG_OPENSSL_ENGINE is not set
CONFIG_PACKAGE_libopenssl-conf=y
```

**Comment by anomeome** [comment]
config that is failing:

```
grep -i OPENSSL configdiff5

CONFIG_LIBCURL_OPENSSL=y
CONFIG_LUA_ECO_OPENSSL=y
CONFIG_OPENSSL_ENGINE=y
CONFIG_OPENSSL_ENGINE_BUILTIN=y
CONFIG_OPENSSL_ENGINE_BUILTIN_AFALG=y
CONFIG_OPENSSL_ENGINE_BUILTIN_DEVCRYPTO=y
CONFIG_OPENSSL_OPTIMIZE_SPEED=y
CONFIG_OPENSSL_PREFER_CHACHA_OVER_GCM=y
CONFIG_OPENSSL_WITH_ASM=y
CONFIG_OPENSSL_WITH_BLAKE2=y
CONFIG_OPENSSL_WITH_CHACHA_POLY1305=y
CONFIG_OPENSSL_WITH_CMS=y
CONFIG_OPENSSL_WITH_DEPRECATED=y
CONFIG_OPENSSL_WITH_DTLS=y
CONFIG_OPENSSL_WITH_EC2M=y
CONFIG_OPENSSL_WITH_ERROR_MESSAGES=y
CONFIG_OPENSSL_WITH_PSK=y
CONFIG_OPENSSL_WITH_SRP=y
CONFIG_OPENSSL_WITH_TLS13=y
CONFIG_OPENVPN_openssl_ENABLE_DEF_AUTH=y
CONFIG_OPENVPN_openssl_ENABLE_FRAGMENT=y
CONFIG_OPENVPN_openssl_ENABLE_IPROUTE2=y
CONFIG_OPENVPN_openssl_ENABLE_LZ4=y
CONFIG_OPENVPN_openssl_ENABLE_LZO=y
CONFIG_OPENVPN_openssl_ENABLE_MULTIHOME=y
CONFIG_OPENVPN_openssl_ENABLE_PF=y
CONFIG_OPENVPN_openssl_ENABLE_PORT_SHARE=y
CONFIG_PACKAGE_libopenssl=y
CONFIG_PACKAGE_libopenssl-conf=y
CONFIG_PACKAGE_libustream-openssl=m
CONFIG_PACKAGE_openssl-util=y
CONFIG_PACKAGE_openvpn-openssl=y
CONFIG_PACKAGE_wpad-basic-openssl=y
# CONFIG_OPENVPN_openssl_ENABLE_SMALL is not set
```
